[Build] Using xnnpack fails with "_Float16 is not supported on this target" for iOS x86_64 simulator

### Describe the issue

When cross-compiling for iOS simulator x86_64, using the `--use_xnnpack` setting causes build to fail. It looks like maybe the build configuration is trying to rely on missing _Float16 support.

Build machine: `M1 MacBook pro, MacOS 15.0`
Xcode: `16.0`
Clang: `16.0.0 (clang-1600.0.26.3)`
Cmake: `3.30.4`
onnxruntime: `c24e55b1f140a2fc3d584738422313d436f122d6`

### Urgency

_No response_

### Target platform

iOS x86_64 (simulator)

### Build script

```
tools/ci_build/github/apple/build_and_assemble_apple_pods.py \
 --include-ops-by-config model.config \
 --build-settings-file build-settings.json
```

build-settings.json:
```
{
  "build_osx_archs": {
    "iphonesimulator": ["x86_64"]
  },
  "build_params": {
    "base": [
      "--parallel",
      "--build_apple_framework",
      "--use_coreml",
      "--skip_tests",
      "--cmake_extra_defines=onnxruntime_BUILD_UNIT_TESTS=OFF"
    ],
    "iphonesimulator": [
      "--ios",
      "--use_xcode",
      "--use_xnnpack",
      "--apple_deploy_target=13.0"
    ]
  }
}
```

### Error / output

```
PrecompileModule /Users/name/code/vendor/onnxruntime/build/apple_framework/intermediates/iphonesimulator_x86_64/Release/build/ExplicitPrecompiledModules/_math-EBU1MF3V4Y9NGBY750RC4B1Q2.scan
    cd /Users/name/code/vendor/onnxruntime/cmake
    builtin-precompileModule /Users/name/code/vendor/onnxruntime/build/apple_framework/intermediates/iphonesimulator_x86_64/Release/build/ExplicitPrecompiledModules/_math-EBU1MF3V4Y9NGBY750RC4B1Q2.scan
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -cc1 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit\=0 -fmessage-length\=295 -ferror-limit 19 -serialize-diagnostic-file /Users/name/code/vendor/onnxruntime/build/apple_framework/intermediates/iphonesimulator_x86_64/Release/build/ExplicitPrecompiledModules/_math-EBU1MF3V4Y9NGBY750RC4B1Q2.dia -fcolor-diagnostics -fsystem-module -fmodule-map-file\=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator18.0.sdk/usr/include/DarwinFoundation.modulemap -fmodule-map-file\=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator18.0.sdk/usr/include/os_availability.modulemap -o /Users/name/code/vendor/onnxruntime/build/apple_framework/intermediates/iphonesimulator_x86_64/Release/build/ExplicitPrecompiledModules/_math-EBU1MF3V4Y9NGBY750RC4B1Q2.pcm -disable-free -emit-module -x c /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/c_standard_library.modulemap -tune-cpu generic -target-cpu core2 -target-feature +sse -target-feature -sse2 -triple x86_64-apple-ios13.0.0-simulator -target-linker-version 1115.7.3 -target-sdk-version\=18.0 -fmodules-validate-system-headers -fno-modules-prune-non-affecting-module-map-files -fno-modulemap-allow-subdirectory-search -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator18.0.sdk -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/16 -fmodule-file\=DarwinFoundation\=/Users/name/code/vendor/onnxruntime/build/apple_framework/intermediates/iphonesimulator_x86_64/Release/build/ExplicitPrecompiledModules/DarwinFoundation-6XBV9A52YVKP0OEUX88WZ9WT8.pcm -fmodule-file\=os_availability\=/Users/name/code/vendor/onnxruntime/build/apple_framework/intermediates/iphonesimulator_x86_64/Release/build/ExplicitPrecompiledModules/os_availability-7M09FJPGLX0VX34H2VI5UFU21.pcm -isystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/16/include -internal-externc-isystem /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator18.0.sdk/usr/include -ivfsstatcache /var/folders/8t/7gch98kn3k15rj57hs17p9900000gn/C/com.apple.DeveloperTools/16.0-16A242d/Xcode/SDKStatCaches.noindex/iphonesimulator18.0-22A3362-db63dc9361471f152f572502bdbfe70a.sdkstatcache -std\=c99 -fasm-blocks -fmodules -fmodule-name\=_math -fmodules-search-all -fno-implicit-modules -fmax-type-align\=16 -mdarwin-stkchk-strong-link -fno-odr-hash-protocols -pic-level 2 -disable-objc-default-synthesize-properties -stack-protector 1 -fgnuc-version\=4.2.1 -fblocks -ffp-contract\=off -fclang-abi-compat\=4.0 -fno-experimental-relative-c++-abi-vtables -fno-file-reproducible -clang-vendor-feature\=+disableNonDependentMemberExprInCurrentInstantiation -clang-vendor-feature\=+enableAggressiveVLAFolding -clang-vendor-feature\=+revert09abecef7bbf -clang-vendor-feature\=+thisNoAlignAttr -clang-vendor-feature\=+thisNoNullAttr -clang-vendor-feature\=+disableAtImportPrivateFrameworkInImplementationError -O2 -fcommon -fno-sanitize-address-use-odr-indicator -fno-sanitize-memory-param-retval -fmerge-all-constants -fno-verbose-asm -fregister-global-dtors-with-atexit -fno-use-init-array -funique-basic-block-section-names -fsplit-dwarf-inlining -gno-codeview-command-line -gstrict-dwarf -gno-column-info -llvm-verify-each -no-struct-path-tbaa -fexperimental-assignment-tracking\=disabled -no-enable-noundef-analysis -stack-protector-buffer-size 0 -finline-functions -fno-unroll-loops -gpubnames -fdiagnostics-hotness-threshold\=0 -fdiagnostics-misexpect-tolerance\=0 -fno-split-cold-code -D CMAKE_INTDIR\=\"Release-iphonesimulator\" -D EIGEN_MPL2_ONLY -D FXDIV_USE_INLINE_ASSEMBLY\=0 -D NDEBUG -D ORT_ENABLE_STREAM -D REDUCED_OPS_BUILD -D XNN_ENABLE_ARM_BF16\=1 -D XNN_ENABLE_ARM_DOTPROD\=1 -D XNN_ENABLE_ARM_FP16_SCALAR\=1 -D XNN_ENABLE_ARM_FP16_VECTOR\=1 -D XNN_ENABLE_ARM_I8MM\=1 -D XNN_ENABLE_ASSEMBLY\=1 -D XNN_ENABLE_AVX256SKX\=1 -D XNN_ENABLE_AVX256VNNI\=1 -D XNN_ENABLE_AVX256VNNIGFNI\=1 -D XNN_ENABLE_AVX512AMX\=1 -D XNN_ENABLE_AVX512FP16\=1 -D XNN_ENABLE_AVX512VNNIGFNI\=1 -D XNN_ENABLE_AVXVNNI\=1 -D XNN_ENABLE_CPUINFO\=1 -D XNN_ENABLE_DWCONV_MULTIPASS\=0 -D XNN_ENABLE_GEMM_M_SPECIALIZATION\=1 -D XNN_ENABLE_HVX\=1 -D XNN_ENABLE_KLEIDIAI\=0 -D XNN_ENABLE_MEMOPT\=1 -D XNN_ENABLE_RISCV_VECTOR\=1 -D XNN_ENABLE_SPARSE\=1 -D XNN_ENABLE_VSX\=1 -D XNN_LOG_LEVEL\=0 -D __GCC_HAVE_DWARF2_CFI_ASM\=1 -MT moduledependenciestarget -dependency-file /Users/name/code/vendor/onnxruntime/build/apple_framework/intermediates/iphonesimulator_x86_64/Release/build/ExplicitPrecompiledModules/_math-EBU1MF3V4Y9NGBY750RC4B1Q2.d -skip-unused-modulemap-deps
While building module '_math':
In file included from <module-includes>:1:
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/math.h:614:27: error: _Float16 is not supported on this target
  614 | extern _Float16 __fabsf16(_Float16) __API_AVAILABLE(macos(15.0), ios(18.0), watchos(11.0), tvos(18.0));
      |                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/math.h:614:8: error: _Float16 is not supported on this target
  614 | extern _Float16 __fabsf16(_Float16) __API_AVAILABLE(macos(15.0), ios(18.0), watchos(11.0), tvos(18.0));
      |        ^
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/math.h:615:28: error: _Float16 is not supported on this target
  615 | extern _Float16 __hypotf16(_Float16, _Float16) __API_AVAILABLE(macos(15.0), ios(18.0), watchos(11.0), tvos(18.0));
      |                            ^

And so on...
```

It might be relevant that: `-D XNN_ENABLE_AVX512FP16\=1`.

### Visual Studio Version

_No response_

### GCC / Compiler Version

_No response_

--- a/.github/workflows/mac.yml
+++ b/.github/workflows/mac.yml
@@ -19,7 +19,7 @@ env:
   python_version: 3.11
 
 jobs:
-  ARM64:
+  ARM64-Xcode16:
     runs-on: macos-14
 
     env:
@@ -59,6 +59,51 @@ jobs:
           --use_coreml \
           --use_xnnpack \
           --use_binskim_compliant_compile_flags
+
+  ARM64-Xcode16-targeting-iphonesimulator-x86_64:
+    runs-on: macos-14
+
+    env:
+      xcode_version: 16
+
+    timeout-minutes: 60
+
+    steps:
+    - uses: actions/setup-python@v5
+      with:
+        python-version: ${{ env.python_version }}
+
+    - name: Verify ARM64 machine
+      shell: python
+      run: |
+        import platform
+        assert platform.machine() == "arm64", "This job expects to be run on an ARM64 machine."
+
+    - name: Use Xcode ${{ env.xcode_version }}
+      shell: bash
+      run: |
+        XCODE_DEVELOPER_DIR="/Applications/Xcode_${{ env.xcode_version }}.app/Contents/Developer"
+        sudo xcode-select --switch "${XCODE_DEVELOPER_DIR}"
+
+    - uses: actions/checkout@v4
+
+    - name: Build
+      shell: bash
+      run: |
+        python ./tools/ci_build/build.py \
+          --build_dir ./build \
+          --update \
+          --build --parallel \
+          --skip_tests \
+          --build_apple_framework \
+          --use_xcode \
+          --use_coreml \
+          --use_xnnpack \
+          --use_binskim_compliant_compile_flags \
+          --ios \
+          --apple_deploy_target=13.0 \
+          --apple_sysroot=iphonesimulator \
+          --osx_arch=x86_64
 
   Vcpkg:
     runs-on: macos-13

--- a/.github/workflows/mac.yml
+++ b/.github/workflows/mac.yml
@@ -87,6 +87,8 @@ jobs:
 
     - uses: actions/checkout@v4
 
+    # Note: Setting onnxruntime_BUILD_UNIT_TESTS=OFF as a workaround for
+    # https://github.com/microsoft/onnxruntime/issues/22245.
     - name: Build
       shell: bash
       run: |
@@ -103,7 +105,8 @@ jobs:
           --ios \
           --apple_deploy_target=13.0 \
           --apple_sysroot=iphonesimulator \
-          --osx_arch=x86_64
+          --osx_arch=x86_64 \
+          --cmake_extra_defines=onnxruntime_BUILD_UNIT_TESTS=OFF
 
   Vcpkg:
     runs-on: macos-13

--- a/cmake/external/xnnpack.cmake
+++ b/cmake/external/xnnpack.cmake
@@ -21,7 +21,12 @@ endif()
 FetchContent_Declare(psimd URL ${DEP_URL_psimd} URL_HASH SHA1=${DEP_SHA1_psimd})
 onnxruntime_fetchcontent_makeavailable(psimd)
 set(PSIMD_SOURCE_DIR ${psimd_SOURCE_DIR})
-FetchContent_Declare(fp16 URL ${DEP_URL_fp16} URL_HASH SHA1=${DEP_SHA1_fp16})
+FetchContent_Declare(
+    fp16
+    URL ${DEP_URL_fp16}
+    URL_HASH SHA1=${DEP_SHA1_fp16}
+    PATCH_COMMAND ${Patch_EXECUTABLE} --binary --ignore-whitespace -p1 < ${PROJECT_SOURCE_DIR}/patches/fp16/remove_math_h_dependency_from_fp16_h.patch
+    )
 onnxruntime_fetchcontent_makeavailable(fp16)
 
 # pthreadpool depends on fxdiv

--- a/cmake/external/xnnpack.cmake
+++ b/cmake/external/xnnpack.cmake
@@ -21,11 +21,37 @@ endif()
 FetchContent_Declare(psimd URL ${DEP_URL_psimd} URL_HASH SHA1=${DEP_SHA1_psimd})
 onnxruntime_fetchcontent_makeavailable(psimd)
 set(PSIMD_SOURCE_DIR ${psimd_SOURCE_DIR})
+
+block(PROPAGATE fp16_PATCH_COMMAND)
+  # only apply fp16 patch for Apple x86_64 targets
+
+  if(APPLE)
+    if(NOT "${CMAKE_OSX_ARCHITECTURES}" STREQUAL "")
+      if ("x86_64" IN_LIST CMAKE_OSX_ARCHITECTURES)
+        set(fp16_PATCH_REQUIRED 1)
+      endif()
+    else()
+      # CMAKE_OSX_ARCHITECTURES unspecified, check host
+      if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+        set(fp16_PATCH_REQUIRED 1)
+      endif()
+    endif()
+  endif()
+
+  if(fp16_PATCH_REQUIRED)
+    message(STATUS "Applying fp16 patch.")
+    set(fp16_PATCH_FILE ${PROJECT_SOURCE_DIR}/patches/fp16/remove_math_h_dependency_from_fp16_h.patch)
+    set(fp16_PATCH_COMMAND ${Patch_EXECUTABLE} --binary --ignore-whitespace -p1 < ${fp16_PATCH_FILE})
+  else()
+    set(fp16_PATCH_COMMAND "")
+  endif()
+endblock()
+
 FetchContent_Declare(
     fp16
     URL ${DEP_URL_fp16}
     URL_HASH SHA1=${DEP_SHA1_fp16}
-    PATCH_COMMAND ${Patch_EXECUTABLE} --binary --ignore-whitespace -p1 < ${PROJECT_SOURCE_DIR}/patches/fp16/remove_math_h_dependency_from_fp16_h.patch
+    PATCH_COMMAND ${fp16_PATCH_COMMAND}
     )
 onnxruntime_fetchcontent_makeavailable(fp16)
 

--- a/cmake/patches/fp16/readme.md
+++ b/cmake/patches/fp16/readme.md
@@ -1,0 +1,11 @@
+# remove_math_h_dependency_from_fp16_h.patch
+
+Remove dependency on math.h (with fabsf()) to work around Xcode 16 build error for iphonesimulator x86_64 target:
+
+/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/math.h:614:27: error:
+      _Float16 is not supported on this target
+  614 | extern _Float16 __fabsf16(_Float16) __API_AVAILABLE(macos(15.0), ios(18.0), watchos(11.0), tvos(18.0));
+      |
+
+This patch was adapted from this PR: https://github.com/Maratyszcza/FP16/pull/32
+See also: https://github.com/google/XNNPACK/issues/6989

--- a/cmake/patches/fp16/remove_math_h_dependency_from_fp16_h.patch
+++ b/cmake/patches/fp16/remove_math_h_dependency_from_fp16_h.patch
@@ -1,0 +1,29 @@
+diff --git a/include/fp16/fp16.h b/include/fp16/fp16.h
+index 2b61fff..1947805 100644
+--- a/include/fp16/fp16.h
++++ b/include/fp16/fp16.h
+@@ -4,10 +4,8 @@
+ 
+ #if defined(__cplusplus) && (__cplusplus >= 201103L)
+ 	#include <cstdint>
+-	#include <cmath>
+ #elif !defined(__OPENCL_VERSION__)
+ 	#include <stdint.h>
+-	#include <math.h>
+ #endif
+ 
+ #ifdef _MSC_VER
+@@ -228,9 +226,11 @@ static inline uint16_t fp16_ieee_from_fp32_value(float f) {
+ 	const float scale_to_inf = fp32_from_bits(UINT32_C(0x77800000));
+ 	const float scale_to_zero = fp32_from_bits(UINT32_C(0x08800000));
+ #endif
+-	float base = (fabsf(f) * scale_to_inf) * scale_to_zero;
+-
+ 	const uint32_t w = fp32_to_bits(f);
++	const float abs_f = fp32_from_bits(w & UINT32_C(0x7FFFFFFF));
++
++	float base = (abs_f * scale_to_inf) * scale_to_zero;
++
+ 	const uint32_t shl1_w = w + w;
+ 	const uint32_t sign = w & UINT32_C(0x80000000);
+ 	uint32_t bias = shl1_w & UINT32_C(0xFF000000);

